# ChatData

## Sqlite
We are using a file system called **sqlite**. It looks and acts like a real single user relational database (RDB). sqlite3 comes packaged with python. You do not need to install the library. You simply import it (as below).

- We are using Pandas to write and read to sqlite. Pandas will manage a lot of the complexity of dealing with a RDB. There are other ways of reading and writing to a RDB that are VERY common and often used in production systems.  

- The most common other way is to read or write to the RDB row by row. As you can see Pandas puts the entire dataframe into the RDB or extracts a new dataframe from the RDB. These are not row by row operations. These are set operations. Set operations ARE more efficient. However, it is common to use row by row operations to avoid needing large memory computers to hold the dataframes. One row only takes a few bytes. An entire dataframe could be many gigabytes and even petabytes of data. Obviously you will end up with problems with your compute resource if your files are this big. This will not happen to you in this activity and may never happen to you while you are a data analyst. Please just be aware of this. 

- This activity is primarily about querying a RDB, so, we are using a simple way that Pandas provides. If you want to learn the more complex way, just google 'reading and wrting to a relational database using python'. There are many resources to learn from.

- The data types that sqlite supports are quite limited. for example, it does not have a DATE type.  This is not a challenge for this project. However, more sophisticated database systems,such as Postgresql, have a large array of data types, such as "DATE", etc. that give those systems additional capability. 

- All of the SQL queries could also be performed on the Pandas DataFrames directly.  You may want to try this yourself for comparison (but make sure you do the SQL queries first, as this is an exercise in using SQL!).


## SQL Magic
Within the Jupyter notebook we will be using something called **SQL Magic**.  This provides a convenient way to write SQL queries directly into code cells in the notebook and to read the results back into a Pandas DataFrame.  This makes working with SQL much easier!

Note that you may need to update your sqlite version using `conda update -c anaconda sqlite` in order for the SQL Magic to work correctly.

## The Data Analysis Lifecycle
The sections in this notebook follow the stages of the Data Analysis Lifecycle introduced in an earlier activity.  The stages are:

- Acquire
- Transform
- Organise
- Analyse
- Communicate
- Maintain


The requirements document for this project is Template SQL queries.xlsx.

# Preliminary Steps: Create a Database
First let's import Sqlite and the other libraries we will need.

In [1]:
# Import Libraries
import numpy as np
import pandas as pd
import sqlite3

### Create the Database
Now we will create the Sqlite database.  Here is some code that does this for you.  We use the `%load_ext` magic command to load the SQL Magic extension and then use `%sql` to connect to the database.

In [2]:
# If the db does not exist, sqlite will create it.
con = sqlite3.connect('chatdata.db')

# loads sql magic
%load_ext sql

# connects sql magic command to the correct db
%sql sqlite:///chatdata.db 

#### Drop the `queries` table if it already exists
The queries table will be our record of the queries created to answer the questions from the requirements spreadsheet.  As we will be running this Jupyter notebook a few times, let's drop (i.e. remove) the queries table so that we start fresh each time.  Here is the code to do this.  We use the `%%sql` magic command to tell Jupyter that we are going to write SQL in the cell.


In [3]:
%%sql
DROP TABLE IF EXISTS 'queries'

 * sqlite:///chatdata.db
Done.


[]

# Task 1: Load the Data
Now we can start loading the data.  The tables will be created as we load the data.  

Note that some of these files are quite large, so make sure you have plenty of free memory!

## Lifecycle Stages: Transform and Organise
The data has already been processed into 3 clean data files ready for this project:

- queries.csv
- posts.csv
- comments.csv

We will load these files into our database.

## Lifecycle Stage: Acquire
The data can be found in the OpenClassrooms instructions for this activity.

### Load Comments Data into a `comments` table
Now we will load the data from the csv files into our sqlite database.

First we load the csv file into a Pandas dataframe:

In [4]:
comments = pd.read_csv('comments.csv')
comments.head()

Id  PostId  Score                                               Text  \
0  723182  385124      0  @BenBolker I don't understand. The fit cannot ...   
1  723183  385124      3  You can't add *less* than (`-min(y)`), but you...   
2  723186  385137      0  nice. If you felt like doing the work it would...   
3  723187  385137      0  i.e. `emdbook::curve3d(-sum(dnbinom(y,mu=mu,si...   
4  723188  385134      0  Don't you mean "so variance should be $\sigma^...   

          CreationDate  UserId  
0  2019-01-01 00:06:39   78575  
1  2019-01-01 00:09:22    2126  
2  2019-01-01 00:32:11    2126  
3  2019-01-01 00:40:36    2126  
4  2019-01-01 00:41:28  112141

Now take the comments dataframe and push the data into the Sqlite database table called 'comments':

In [5]:
# load comments into sqlite
#This will make a SQL table called 'comments' based on the comments dataframe, if there is one with the same name replace
# and do not keep the dataframe index in the new sql table
comments.to_sql('comments', con, if_exists='replace', index=False)

# read back in to prove that it worked
#This turns a SQL query into a variable 'sql', the con has already been specified and is used to help the pd.read_sql function
#find the location and details of the database it will pull data from. the function copies the contents of the SQL table into a pandas DataFrame, 
#so that you can manipulate and analyze the data more easily.
sql = 'SELECT * FROM comments'
comments = pd.read_sql(sql, con)
comments.head()

Id  PostId  Score                                               Text  \
0  723182  385124      0  @BenBolker I don't understand. The fit cannot ...   
1  723183  385124      3  You can't add *less* than (`-min(y)`), but you...   
2  723186  385137      0  nice. If you felt like doing the work it would...   
3  723187  385137      0  i.e. `emdbook::curve3d(-sum(dnbinom(y,mu=mu,si...   
4  723188  385134      0  Don't you mean "so variance should be $\sigma^...   

          CreationDate  UserId  
0  2019-01-01 00:06:39   78575  
1  2019-01-01 00:09:22    2126  
2  2019-01-01 00:32:11    2126  
3  2019-01-01 00:40:36    2126  
4  2019-01-01 00:41:28  112141

### Load Other CSVs
Now that you have seen the code for loading in the comments.csv now do the same to read and write the posts.csv and users.csv to sqlite. 

<div style="background-color:lemonchiffon; border:1px solid; padding:5px; margin:5px; font-weight:bold"><font color="crimson">
TODO: Enter code in the following cells. Insert as many cells as you want to do this.
    </font></div>

In [6]:
#loading posts.csv and creating pandas dataframe
posts = pd.read_csv('posts.csv')
#creating SQL table
posts.to_sql('posts', con, if_exists='replace', index=False)
#checking the new posts table exists
sql2 = 'SELECT * FROM posts'
posts = pd.read_sql(sql2, con)
posts.head(3)

Id  PostTypeId  AcceptedAnswerId  ParentId         CreationDate  Score  \
0  423497           1            423511         0  2019-08-24 09:39:31      2   
1  423498           1                 0         0  2019-08-24 09:47:42      1   
2  423499           1                 0         0  2019-08-24 09:48:26      1   

   ViewCount                                               Body  OwnerUserId  \
0         68  <p>From wikipedia <a href="https://en.wikipedi...        64552   
1         24  <p>I am currently doing local sensitivity anal...        87231   
2         56  <p>I'm an honours student in psychology doing ...       257207   

  OwnerDisplayName  ...  LastEditorDisplayName         LastEditDate  \
0             None  ...                   None                 None   
1             None  ...                   None  2019-09-06 12:52:32   
2             None  ...                   None                 None   

      LastActivityDate                                              Title  \
0  2019-08-24 11:38:54  When are biased estimators with lower MSE pref...   
1  2019-09-06 12:52:32  How to interpret the result from local sensiti...   
2  2019-08-25 08:54:17      Power Analysis in G-Power - Mixed Model Anova   

                                 Tags AnswerCount  CommentCount  \
0     <bias><unbiased-estimator><mse>           1             0   
1  <sensitivity-analysis><elasticity>           1             0   
2                     <anova><gpower>           2             0   

   FavoriteCount           ClosedDate CommunityOwnedDate  
0              1  2019-08-25 10:25:24               None  
1              0                 None               None  
2              0                 None               None  

[3 rows x 21 columns]

In [7]:
#loading users.csv and creating pandas dataframe
users = pd.read_csv('users.csv')
#creating SQL table
users.to_sql('users', con, if_exists='replace', index=False)
#checking the new posts table exists
sql3 = 'SELECT * FROM users'
users = pd.read_sql(sql3, con)
users.head(3)

Id  Reputation         CreationDate DisplayName       LastAccessDate  \
0  157607          31  2017-04-17 14:50:42  user157607  2019-07-23 16:44:08   
1  157656         101  2017-04-17 20:08:20  user102859  2019-06-26 13:42:13   
2  157704         133  2017-04-18 05:10:47      jupiar  2019-11-25 13:32:27   

  WebsiteUrl         Location  \
0       None             None   
1       None             None   
2       None  Shanghai, China   

                                             AboutMe  Views  UpVotes  \
0                                               None      0        0   
1                                               None      3        0   
2  <p>Originally from the U.K, I have an Undergra...      1        1   

   DownVotes                                    ProfileImageUrl  AccountId  
0          0  https://www.gravatar.com/avatar/2efb161849efa4...   10705183  
1          0      https://i.stack.imgur.com/eY4ka.jpg?s=128&g=1   10567606  
2          0  https://www.gravatar.com/avatar/720e20205122c5...    9501631

### **TODO**: Drop Duplicates
Look for and drop any duplicates in all 3 of the tables (if they exist). Use Pandas to do this. If you find duplicates, you will need to rewrite the table.

<div style="background-color:lemonchiffon; border:1px solid; padding:5px; margin:5px; font-weight:bold"><font color="crimson">
TODO: Enter code in the following cells. 
    </font></div>

In [8]:
#Looking for duplicates in comments table
comments.Id.duplicated(keep = False).sum()

0

In [9]:
#Looking for duplicates in posts table
posts.Id.duplicated(keep = False).sum()

0

In [10]:
#Looking for duplicates in users table
users.Id.duplicated(keep = False).sum()

0

## Review the Data

Spend some time reviewing the data.  Understand what data we have, think about how that data can be used to assist in the initiative of understanding how ChatData is used in the real world.  Is the data organised in a way that would lend itself to being manged in a relational database?  How would the different tables be connected?  What are the primary and foreign keys?  Would this give you a 3NF model?

Also think about security and ethics.  Is there personal data in here?  Could individuals be identified through this data?  It it ethical to use the data in this way?  You will be asked to comment on these questions later!

Use the code below to help you.

In [11]:
users.columns

Index(['Id', 'Reputation', 'CreationDate', 'DisplayName', 'LastAccessDate',
       'WebsiteUrl', 'Location', 'AboutMe', 'Views', 'UpVotes', 'DownVotes',
       'ProfileImageUrl', 'AccountId'],
      dtype='object')

In [12]:
users.head()

Id  Reputation         CreationDate DisplayName       LastAccessDate  \
0  157607          31  2017-04-17 14:50:42  user157607  2019-07-23 16:44:08   
1  157656         101  2017-04-17 20:08:20  user102859  2019-06-26 13:42:13   
2  157704         133  2017-04-18 05:10:47      jupiar  2019-11-25 13:32:27   
3  157709         155  2017-04-18 06:39:18      farmer  2019-02-17 19:44:24   
4  157755         101  2017-04-18 12:56:17      Miki P  2019-08-12 17:02:21   

  WebsiteUrl         Location  \
0       None             None   
1       None             None   
2       None  Shanghai, China   
3       None             None   
4       None             None   

                                             AboutMe  Views  UpVotes  \
0                                               None      0        0   
1                                               None      3        0   
2  <p>Originally from the U.K, I have an Undergra...      1        1   
3                                               None     16        0   
4                                               None      1        9   

   DownVotes                                    ProfileImageUrl  AccountId  
0          0  https://www.gravatar.com/avatar/2efb161849efa4...   10705183  
1          0      https://i.stack.imgur.com/eY4ka.jpg?s=128&g=1   10567606  
2          0  https://www.gravatar.com/avatar/720e20205122c5...    9501631  
3          0  https://www.gravatar.com/avatar/0f8c4bde3d8f25...   10709837  
4          0  https://www.gravatar.com/avatar/af088558cd81c5...    7969290

In [13]:
comments.columns

Index(['Id', 'PostId', 'Score', 'Text', 'CreationDate', 'UserId'], dtype='object')

In [14]:
comments.head()

Id  PostId  Score                                               Text  \
0  723182  385124      0  @BenBolker I don't understand. The fit cannot ...   
1  723183  385124      3  You can't add *less* than (`-min(y)`), but you...   
2  723186  385137      0  nice. If you felt like doing the work it would...   
3  723187  385137      0  i.e. `emdbook::curve3d(-sum(dnbinom(y,mu=mu,si...   
4  723188  385134      0  Don't you mean "so variance should be $\sigma^...   

          CreationDate  UserId  
0  2019-01-01 00:06:39   78575  
1  2019-01-01 00:09:22    2126  
2  2019-01-01 00:32:11    2126  
3  2019-01-01 00:40:36    2126  
4  2019-01-01 00:41:28  112141

In [15]:
posts.columns

Index(['Id', 'PostTypeId', 'AcceptedAnswerId', 'ParentId', 'CreationDate',
       'Score', 'ViewCount', 'Body', 'OwnerUserId', 'OwnerDisplayName',
       'LastEditorUserId', 'LastEditorDisplayName', 'LastEditDate',
       'LastActivityDate', 'Title', 'Tags', 'AnswerCount', 'CommentCount',
       'FavoriteCount', 'ClosedDate', 'CommunityOwnedDate'],
      dtype='object')

In [16]:
posts.head()

Id  PostTypeId  AcceptedAnswerId  ParentId         CreationDate  Score  \
0  423497           1            423511         0  2019-08-24 09:39:31      2   
1  423498           1                 0         0  2019-08-24 09:47:42      1   
2  423499           1                 0         0  2019-08-24 09:48:26      1   
3  423500           2                 0    215865  2019-08-24 09:57:01      0   
4  423502           2                 0    423286  2019-08-24 10:44:52      3   

   ViewCount                                               Body  OwnerUserId  \
0         68  <p>From wikipedia <a href="https://en.wikipedi...        64552   
1         24  <p>I am currently doing local sensitivity anal...        87231   
2         56  <p>I'm an honours student in psychology doing ...       257207   
3          0  <p>Maybe you can look this <a href="https://me...       106606   
4          0  <blockquote>\n  <p>Q1) Is my approach valid?</...       220643   

  OwnerDisplayName  ...  LastEditorDisplayName         LastEditDate  \
0             None  ...                   None                 None   
1             None  ...                   None  2019-09-06 12:52:32   
2             None  ...                   None                 None   
3             None  ...                   None                 None   
4             None  ...                   None  2019-08-24 12:13:46   

      LastActivityDate                                              Title  \
0  2019-08-24 11:38:54  When are biased estimators with lower MSE pref...   
1  2019-09-06 12:52:32  How to interpret the result from local sensiti...   
2  2019-08-25 08:54:17      Power Analysis in G-Power - Mixed Model Anova   
3  2019-08-24 09:57:01                                               None   
4  2019-08-24 12:13:46                                               None   

                                 Tags AnswerCount  CommentCount  \
0     <bias><unbiased-estimator><mse>           1             0   
1  <sensitivity-analysis><elasticity>           1             0   
2                     <anova><gpower>           2             0   
3                                None           0             0   
4                                None           0             1   

   FavoriteCount           ClosedDate CommunityOwnedDate  
0              1  2019-08-25 10:25:24               None  
1              0                 None               None  
2              0                 None               None  
3              0                 None               None  
4              0                 None               None  

[5 rows x 21 columns]

# Working with Sqlite and SQL Magic
In this section let's spend a little time understanding a bit more about how we can work with Sqlite within Jupyter.

Let's look at 2 ways to query the sqlite database: using SQL Magic or using Pandas.  Either way is fine for this project.

## Writing queries with SQL Magic

You will now need to write some queries to get answers to the questions in the requirements.  

For single-line queries, start the cell with `%sql` and simply enter your query:

In [17]:
# This is an example
%sql SELECT COUNT(*) FROM comments

 * sqlite:///chatdata.db
Done.


[(50000,)]

For multi line sql statements use `%%sql` as follows. This tells Jupyter that *everything* in this cell should be interpreted as sql. So, NO comments other statements are allowed:

In [18]:
%%sql
SELECT Id, PostId, Score, Text
    FROM comments
        LIMIT 5

 * sqlite:///chatdata.db
Done.


[(723182, 385124, 0, '@BenBolker I don\'t understand. The fit cannot be done for the negative $y$. So intuitively I\'d think that in order to retain the relativity of the  ... (24 characters truncated) ... rror" or "flip" it to the positive axis? Would it be possible to adjust every point individually? Basically e.g. take $abs(y_i)$ instead of minimums.'),
 (723183, 385124, 3, "You can't add *less* than (`-min(y)`), but you could add *more*. I'm going to stop answering now sorry, because **judging what the 'best' approach is ... (60 characters truncated) ...  of the analysis, why you need to fit an exponential, your level of computational and statistical sophistication and that of your audience, etc. ...)"),
 (723186, 385137, 0, 'nice. If you felt like doing the work it would be nice to generate an image/contour plot of log-likelihood as a function of (mu, theta) and show the lines corresponding to the two `size` values .'),
 (723187, 385137, 0, 'i.e. `emdbook::curve3d(-sum(dnbinom(y,mu=mu,size=size,log=TRUE)),\n                 xlim=c(2,5),ylim=c(0.2,0.5),\n                 varnames=c("mu","s ... (57 characters truncated) ... ,col="red")\nabline(h=fit.what$theta,col="blue")\n` ... although doesn\'t look this is actually the answer - mu-hat is independent of theta-hat ... ?'),
 (723188, 385134, 0, 'Don\'t you mean "so variance should be $\\sigma^2/(n\\mu^2)$"')]

## Writing queries with Pandas
Another way to write queries is to use pandas:

In [19]:
sql = """
SELECT Id, PostId, Score, Text
    FROM comments
        LIMIT 5
        """
result = pd.read_sql(sql, con)
result

Id  PostId  Score                                               Text
0  723182  385124      0  @BenBolker I don't understand. The fit cannot ...
1  723183  385124      3  You can't add *less* than (`-min(y)`), but you...
2  723186  385137      0  nice. If you felt like doing the work it would...
3  723187  385137      0  i.e. `emdbook::curve3d(-sum(dnbinom(y,mu=mu,si...
4  723188  385134      0  Don't you mean "so variance should be $\sigma^...

# Creating Tables with Referential Integrity

When we loaded the csv files into Sqlite database tables, Sqlite created the tables for us behind the scenes. Let's inspect this a bit more.

We can see how Sqlite created the tables by querying the `sqlite_master` table, which Sqlite uses to keep track of what objects have been created in the database:

In [20]:
%%sql
select sql from sqlite_master

 * sqlite:///chatdata.db
Done.


[('CREATE TABLE "comments" (\n"Id" INTEGER,\n  "PostId" INTEGER,\n  "Score" INTEGER,\n  "Text" TEXT,\n  "CreationDate" TEXT,\n  "UserId" INTEGER\n)',),
 ('CREATE TABLE "posts" (\n"Id" INTEGER,\n  "PostTypeId" INTEGER,\n  "AcceptedAnswerId" INTEGER,\n  "ParentId" INTEGER,\n  "CreationDate" TEXT,\n  "Scor ... (240 characters truncated) ...   "Tags" TEXT,\n  "AnswerCount" INTEGER,\n  "CommentCount" INTEGER,\n  "FavoriteCount" INTEGER,\n  "ClosedDate" TEXT,\n  "CommunityOwnedDate" TEXT\n)',),
 ('CREATE TABLE "users" (\n"Id" INTEGER,\n  "Reputation" INTEGER,\n  "CreationDate" TEXT,\n  "DisplayName" TEXT,\n  "LastAccessDate" TEXT,\n  "WebsiteUr ... (17 characters truncated) ... tion" TEXT,\n  "AboutMe" TEXT,\n  "Views" INTEGER,\n  "UpVotes" INTEGER,\n  "DownVotes" INTEGER,\n  "ProfileImageUrl" TEXT,\n  "AccountId" INTEGER\n)',)]

**When a table or view is created the SQL code used to define it is saved in a system table called sqlite_master. The code above checks what has been saved. The select sql from sqlite_master command queries this system table to retrieve the SQL code that was used to create each table or view in the database. By looking at this SQL code, you can understand how the tables or views are defined and what columns they contain, among other things.** 

The above results show the `CREATE TABLE` statements that could be used by Sqlite to recreate the tables with the exact same structure.

The problem with the CREATE TABLE statements above is that they don't enforce **referential integrity**.  In other words, they don't ensure that every UserId and PostId in the comments table refers to an actual UserId and PostId in the users and posts tables.  At the moment. we can insert any old number here, and even have multiple users with the same Id!  One of the advantages of working with relational databases is that they can enforce the correct uniqueness and relationships in the data, but at the moment we are not using that feature.  So let's fix that...

First, let's drop the original tables:


In [21]:
%%sql
DROP TABLE comments;
DROP TABLE users;
DROP TABLE posts;

 * sqlite:///chatdata.db
Done.
Done.
Done.


[]

**The DROP TABLE statement is used to drop tables in SQL, and the code is using this statement to drop three tables: comments, users, and posts. This means that any data and metadata associated with these tables will be deleted from the database.**

Prove that this worked by selecting the names of the tables back.  We should have no tables:

In [22]:
%%sql
SELECT name FROM sqlite_master WHERE type='table'
    ORDER BY name

 * sqlite:///chatdata.db
Done.


[]

**The SELECT statement is used to retrieve data from the database. In this case, to retrieve the names of the tables in the database.**

**The FROM clause specifies the table or tables to retrieve data from. In this case, the sqlite_master table, a system table in SQLite that contains metadata about all the objects in the database.**

**The WHERE clause is used to filter the results of the query. In this case, using the type column of the sqlite_master table to filter for objects of type 'table'.**

**The ORDER BY clause is used to sort the results of the query in ascending order by the name column.**

In Sqlite we need to enable the enforcement of foreign key constraints foreign key constraints are not enabled by default. Therefore, you need to explicitly enable them by running the PRAGMA foreign_keys=ON; statement before you can create tables with foreign key constraints or insert data into tables that have foreign keys:

In [23]:
%%sql
PRAGMA foreign_keys=ON;

 * sqlite:///chatdata.db
Done.


[]

Now recreate the users table with a **primary key constraint** by copying the CREATE TABLE statement from above and adding the NOT NULL PRIMARY KEY clause to the Id:

In [24]:
%%sql
CREATE TABLE "users" (
    "Id" INTEGER NOT NULL PRIMARY KEY,
    "Reputation" INTEGER,
    "CreationDate" TEXT,
    "DisplayName" TEXT,
    "LastAccessDate" TEXT,
    "WebsiteUrl" TEXT,
    "Location" TEXT,
    "AboutMe" TEXT,
    "Views" INTEGER,
    "UpVotes" INTEGER,
    "DownVotes" INTEGER,
    "ProfileImageUrl" TEXT,
    "AccountId" INTEGER
);


 * sqlite:///chatdata.db
Done.


[]

Now do the same for the posts table:

<div style="background-color:lemonchiffon; border:1px solid; padding:5px; margin:5px; font-weight:bold"><font color="crimson">
TODO: Complete the following code cell
    </font></div>

Run the CREATE TABLE statement for the posts table, including the primary key constraint

In [25]:
%%sql
CREATE TABLE "posts" (
    "Id" INTEGER NOT NULL PRIMARY KEY,
    "PostTypeId" INTEGER,
    "AcceptedAnswerId" INTEGER,
    "ParentId" INTEGER,
    "CreationDate" TEXT,
    "Score" INTEGER,
    "ViewCount" INTEGER,
    "Body" TEXT,
    "OwnerUserId" INTEGER,
    "OwnerDisplayName" TEXT,
    "LastEditorUserId" INTEGER,
    "LastEditorDisplayName" TEXT,
    "LastEditDate" TEXT,
    "LastActivityDate" TEXT,
    "Title" TEXT,
    "Tags" TEXT,
    "AnswerCount" INTEGER,
    "CommentCount" INTEGER,
    "FavoriteCount" INTEGER,
    "ClosedDate" TEXT,
    "CommunityOwnedDate" TEXT,
     FOREIGN KEY("OwnerUserId") REFERENCES "users"("Id")
);


 * sqlite:///chatdata.db
Done.


[]

Now for the comments table.  We need to add the primary key constraint on the id here as we did for users and posts, but we also need to add FOREIGN KEY constraints on the UserId and PostId. Read the documentation here and find our how to do that: https://www.sqlite.org/foreignkeys.html.  Then create the comments table with the correct constraints:

<div style="background-color:lemonchiffon; border:1px solid; padding:5px; margin:5px; font-weight:bold"><font color="crimson">
TODO: Complete the following code cell
    </font></div>

In [26]:
# Run the CREATE TABLE statement for the comments table, including the primary key constraint and the 2 foreign key constraints
# TODO

In [27]:
%%sql
CREATE TABLE "comments" (
    "Id" INTEGER NOT NULL PRIMARY KEY,
    "PostId" INTEGER,
    "Score" INTEGER,
    "Text" TEXT,
    "CreationDate" TEXT,
    "UserId" INTEGER,
    FOREIGN KEY("PostId") REFERENCES "posts"("Id"),
    FOREIGN KEY("UserId") REFERENCES "users"("Id")
);

 * sqlite:///chatdata.db
Done.


[]

Now we can re-insert the data into these constrained tables.  First users:

In [28]:
# Insert data into the new users table
users.to_sql('users', con, if_exists='append', index=False)

Now posts:

<div style="background-color:lemonchiffon; border:1px solid; padding:5px; margin:5px; font-weight:bold"><font color="crimson">
TODO: Complete the following code cell
    </font></div>

In [29]:
# Insert data into the new posts table
posts.to_sql('posts', con, if_exists='append', index=False)

Finally comments, which references the users and posts tables:

<div style="background-color:lemonchiffon; border:1px solid; padding:5px; margin:5px; font-weight:bold"><font color="crimson">
TODO: Complete the following code cell
    </font></div>

In [30]:
# Insert data into the new comments table
comments.to_sql('comments', con, if_exists='append', index=False)

Now check that we have the 3 new table definitions in Sqlite:

In [31]:
%%sql
SELECT name FROM sqlite_master WHERE type='table'
    ORDER BY name

 * sqlite:///chatdata.db
Done.


[('comments',), ('posts',), ('users',)]

We now have all the data in tables in Sqlite and the tables will enforce the referential integrity.

# Example Query and Pattern for Tasks 2 and 3

As you work through the next tasks, you will need to:

1. Prepare the Sqlite query to answer the question
2. Test it
3. Insert it into the `queries` table, so we have a record of it for others.

This is the process that we want you to follow for this project while completing it. 

Let's see an example of this by answering the following question:

**Which 5 users have viewed the most times and what is the sum of those views per user?**

## Prepare the Sqlite query

First, let's write the query:

In [32]:
sql = """
SELECT Id, SUM(Views) AS TotalViews
    FROM Users
        GROUP BY Id
            ORDER BY TotalViews DESC
                LIMIT 5
                """
result = pd.read_sql(sql, con) # con is the connection to the database
result

Id  TotalViews
0   919       85180
1  4253       35119
2   805       34637
3  7290       32639
4  3277       29255

## Test the query
You can optionally prove the query worked by performing the same query in Pandas:

In [33]:
results = users.groupby(['Id']).sum().sort_values('Views', ascending = False)[:5]
results['Views']

Id
919     85180
4253    35119
805     34637
7290    32639
3277    29255
Name: Views, dtype: int64

## Insert the query into the `queries` table
Now we need to put this query into the `queries` table in sqlite. Remember we want these queries to be accessible to everybody that should have access to them. We do not want people writing and rewriting the same queires over and over again. The easiest thing to do is create a dictionary with the values and insert these into the queries table.  Note that the values are provided as lists as we are inserting a list of values (i.e. a number of rows) into the table.  In this case the number of rows is 1, so we have lists of 1 item.

So here, we have a column called 'task' with a list of values, a column called 'action' with a list of values, etc.

In [34]:
query_dict = {
              'task': ['Single Table Queries'],
              'action': ['Which 5 users have viewed the most times and what is the sum of those views per user?'],
              'query': [sql]
             }
query_dict

{'task': ['Single Table Queries'],
 'action': ['Which 5 users have viewed the most times and what is the sum of those views per user?'],
 'query': ['\nSELECT Id, SUM(Views) AS TotalViews\n    FROM Users\n        GROUP BY Id\n            ORDER BY TotalViews DESC\n                LIMIT 5\n                ']}

Now that you have the data structure (query_dict) containing the data, create a pandas dataframe that holds those values:

In [35]:
queries = pd.DataFrame(query_dict)
queries

task                                             action  \
0  Single Table Queries  Which 5 users have viewed the most times and w...   

                                               query  
0  \nSELECT Id, SUM(Views) AS TotalViews\n    FRO...

Now load that pandas dataframe (queries) into the sqlite table called queries. In this case, you use append NOT replace. You will be adding to this tables as you go thru this project.

In [36]:
# load query into sqlite
queries.to_sql('queries', con, if_exists='append', index=False)

# read back in to prove that it worked
sql = 'SELECT * FROM queries'
queries = pd.read_sql(sql, con)
queries.head()

task                                             action  \
0  Single Table Queries  Which 5 users have viewed the most times and w...   

                                               query  
0  \nSELECT Id, SUM(Views) AS TotalViews\n    FRO...

It is likely that as you iterate thru this notebook you will create some duplicate entries in the query table. Not a big deal. Just tell sql or pandas (much easier in pandas) to drop the duplicates! If you drop the duplicates in Pandas you will have to write the entire dataframe back to sqlite. Otherwise it is changed in memory in Pandas but NOT in sqlite on the disk drive (or SDD). Just do this drop at the end of this notebook so that you are not constantly dealing with this. 

So, to summarise, as you go through the following tasks you need to:

- answer the question in sql
- prove it in pandas (if you want to)
- put the query into the queries table

# Task 1 (continued): Insert the CREATE TABLE Statements into the `queries` Table

Now that we understand how to populate the `queries` table, let's insert the CREATE TABLE statements into it.  First let's define a function to help us insert into the queries table:


In [37]:
# Define a function that will insert into the queries table
def store_query(task, action, query):
    query_dict = {
                  'task': [task],
                  'action': [action],
                  'query': [query]
                 }

    # put query into the query_dict
    queries = pd.DataFrame(query_dict)

    # load query into sqlite
    queries.to_sql('queries', con, if_exists='append', index=False)    

Now we can specify the queries and call the above function to store them.  The first one is done for you.

In [38]:
sql = """
    CREATE TABLE "comments" (
    "Id" INTEGER,
    "PostId" INTEGER,
    "Score" INTEGER,
    "Text" TEXT,
    "CreationDate" TEXT,
    "UserId" INTEGER
    )
    """
store_query("Creating Tables", "Create table comments", sql)

Let's prove it works by selecting back from the queries table:

In [39]:
# Prove it works
%sql SELECT * FROM queries

 * sqlite:///chatdata.db
Done.


[('Single Table Queries', 'Which 5 users have viewed the most times and what is the sum of those views per user?', '\nSELECT Id, SUM(Views) AS TotalViews\n    FROM Users\n        GROUP BY Id\n            ORDER BY TotalViews DESC\n                LIMIT 5\n                '),
 ('Creating Tables', 'Create table comments', '\n    CREATE TABLE "comments" (\n    "Id" INTEGER,\n    "PostId" INTEGER,\n    "Score" INTEGER,\n    "Text" TEXT,\n    "CreationDate" TEXT,\n    "UserId" INTEGER\n    )\n    ')]

## Insert the other CREATE TABLE statements into the `queries` table.
Follow the above pattern to complete these code cells:

<div style="background-color:lemonchiffon; border:1px solid; padding:5px; margin:5px; font-weight:bold"><font color="crimson">
TODO: Complete the following code cells
    </font></div>

In [40]:
# Insert the CREATE TABLE for posts into the queries table
sql2 = """
    CREATE TABLE "posts" (
    "Id" INTEGER NOT NULL PRIMARY KEY,
    "PostTypeId" INTEGER,
    "AcceptedAnswerId" INTEGER,
    "ParentId" INTEGER,
    "CreationDate" TEXT,
    "Score" INTEGER,
    "ViewCount" INTEGER,
    "Body" TEXT,
    "OwnerUserId" INTEGER,
    "OwnerDisplayName" TEXT,
    "LastEditorUserId" INTEGER,
    "LastEditorDisplayName" TEXT,
    "LastEditDate" TEXT,
    "LastActivityDate" TEXT,
    "Title" TEXT,
    "Tags" TEXT,
    "AnswerCount" INTEGER,
    "CommentCount" INTEGER,
    "FavoriteCount" INTEGER,
    "ClosedDate" TEXT,
    "CommunityOwnedDate" TEXT
    )
    """
store_query("Creating Tables", "Create table posts", sql2)

In [41]:
# Insert the CREATE TABLE for users into the queries table
sql3 = """
    CREATE TABLE "users" (
    "Id" INTEGER NOT NULL PRIMARY KEY,
    "Reputation" INTEGER,
    "CreationDate" TEXT,
    "DisplayName" TEXT,
    "LastAccessDate" TEXT,
    "WebsiteUrl" TEXT,
    "Location" TEXT,
    "AboutMe" TEXT,
    "Views" INTEGER,
    "UpVotes" INTEGER,
    "DownVotes" INTEGER,
    "ProfileImageUrl" TEXT,
    "AccountId" INTEGER
    );
    """
store_query("Creating Tables", "Create table users", sql3)

## Count the Number of Rows in Each Table
Run some queries to count the number of rows in each of the tables.  Don't forget to insert the query into the queries table.

<div style="background-color:lemonchiffon; border:1px solid; padding:5px; margin:5px; font-weight:bold"><font color="crimson">
TODO: Complete the following code cells
    </font></div>

In [42]:
# Count the number of rows in the comments table
len(comments)

50000

In [43]:
# convert the tuple to a DataFrame
comments_df = pd.DataFrame(comments, columns=['Id', 'PostId', 'Score','Text','CreationDate','UserId'])

# display the DataFrame
comments_df.head()

Id  PostId  Score                                               Text  \
0  723182  385124      0  @BenBolker I don't understand. The fit cannot ...   
1  723183  385124      3  You can't add *less* than (`-min(y)`), but you...   
2  723186  385137      0  nice. If you felt like doing the work it would...   
3  723187  385137      0  i.e. `emdbook::curve3d(-sum(dnbinom(y,mu=mu,si...   
4  723188  385134      0  Don't you mean "so variance should be $\sigma^...   

          CreationDate  UserId  
0  2019-01-01 00:06:39   78575  
1  2019-01-01 00:09:22    2126  
2  2019-01-01 00:32:11    2126  
3  2019-01-01 00:40:36    2126  
4  2019-01-01 00:41:28  112141

In [44]:
# Count the number of rows in the comments table
#the [0] next to shape specifies the number of rows
print('The number of rows in the comments table are:',comments.shape[0])

The number of rows in the comments table are: 50000


In [45]:
%%sql
SELECT COUNT (*)
FROM posts;

 * sqlite:///chatdata.db
Done.


[(42234,)]

In [46]:
%%sql
SELECT COUNT (*)
FROM users;

 * sqlite:///chatdata.db
Done.


[(18412,)]

## Do some Random Checks on the Data
Let's write some queries that select 5 random rows from each table.  The queries are provided here:

    select * from Comments order by random() limit 5;
    select * from Users order by random() limit 5;
    select * from Posts order by random() limit 5;

Enter the queries into the code cells below and insert the queries into the `queries` table in the same was as you did for the CREATE TABLE statements.

<div style="background-color:lemonchiffon; border:1px solid; padding:5px; margin:5px; font-weight:bold"><font color="crimson">
TODO: Complete the following code cells
    </font></div>

In [47]:
# Run the query to select 5 random rows from the posts table
# TODO

In [48]:
%%sql
select * from Comments order by random() limit 5;

 * sqlite:///chatdata.db
Done.


[(746416, 398426, 0, 'In SPSS, there is a classification table. Yes, I reduced the cut-off from 0.5 down to 0.10 and eventually zeroed out the typeII error.\nI am trying t ... (85 characters truncated) ... n make a significantly better prediction. Also, I am trying to figure out from a list of many risk factors, which are more related to this diagnosis.', '2019-03-20 00:16:53', 241742),
 (768804, 411719, 0, '@djlid Yes, I agree. Of course, the 100% is used by propagandists, even though it is completely meaningless.', '2019-06-05 23:17:30', 55548),
 (752897, 402319, 1, 'Generally speaking, using the lowest sum of (absolute value of error) is more robust to outliers that lowest sum of (error squared). Ann outlier with an error value of 5 becomes 25 when squared.', '2019-04-10 22:13:43', 35584),
 (791660, 423408, 0, "why don't u invite me to chat session so we can continue this , if you can't do this we can communicate via email", '2019-08-29 14:19:39', 3382),
 (785137, 420681, 0, 'Thanks, it took me some time to process, but it is a very detailed explanation of how to tackle the problem. Mathematics is such a beautiful language. However, I am not quite sure how to read the infinity sign in equation (1). I should brush up on my math, but maybe you like to explain?', '2019-08-06 11:44:06', 200983)]

In [49]:
# Run the query to select 5 random rows from the posts comments
# TODO

In [50]:
%%sql
select * from Users order by random() limit 5

 * sqlite:///chatdata.db
Done.


[(142801, 141, '2016-12-18 20:03:39', 'Anthony', '2019-10-23 17:59:29', None, None, None, 0, 0, 0, 'https://www.gravatar.com/avatar/2f3ba4237a6edc799533c393afafa09e?s=128&d=identicon&r=PG&f=1', 5642665),
 (48973, 185, '2014-06-25 10:53:13', 'r.bot', '2019-11-27 16:37:17', None, None, '<p></p>\n', 5, 41, 0, 'https://i.stack.imgur.com/TSH6j.jpg?s=128&g=1', 3112082),
 (244039, 11, '2019-04-08 19:19:42', 'Nils Rudqvist', '2019-10-18 14:27:58', None, None, None, 0, 0, 0, 'https://lh5.googleusercontent.com/-EKMfMx07Omc/AAAAAAAAAAI/AAAAAAAAeHE/Asb7-jftY70/photo.jpg?sz=128', 11126770),
 (260535, 15, '2019-09-22 12:37:35', 'Timmy', '2019-11-15 15:24:46', None, None, None, 4, 0, 0, 'https://www.gravatar.com/avatar/842de385be190714df40f6c7fa638463?s=128&d=identicon&r=PG', 12634996),
 (142896, 179, '2016-12-19 18:35:06', 'Mustard Tiger', '2019-10-28 02:58:31', None, 'Canada', None, 9, 39, 0, 'https://i.stack.imgur.com/dCKEb.jpg', 5496408)]

In [51]:
# Run the query to select 5 random rows from the users table
# TODO

In [52]:
%%sql
select * from Posts order by random() limit 5

 * sqlite:///chatdata.db
Done.


[(390247, 1, 0, 0, '2019-02-01 00:16:50', 1, 65, '<p>Let\'s say I\'ve got panel data where a count outcome <span class="math-container">$y$</span> and continuous independent variable <span class="mat ... (1002 characters truncated) ...  something like this in STATA:</p>\n\n<p>"poisson ybar x, vce(cluster i)"</p>\n\n<p>if ybar=<span class="math-container">$\\overline{y}$</span></p>\n', 181146, None, 0, None, None, '2019-02-01 00:16:50', 'Are Poisson Regressions with Serial Correlation Biased or Inconsistent? (No Fixed Effects)', '<autocorrelation><bias><poisson-regression><consistency><clustered-standard-errors>', 0, 4, 0, None, None),
 (408964, 2, 0, 408955, '2019-05-18 17:16:41', 0, 0, '<p>It is difficult to judge normality with only <span class="math-container">$n = 6$</span> observations. Many small samples, regardless of origin, w ... (874 characters truncated) ... lue)\nmean(pv.u &gt; .05)\n## 0.93769\npv.e = replicate(10^5,shapiro.test(rexp(6))$</span>p.value)\nmean(pv.e &gt; .05)\n[1] 0.78049\n</code></pre>\n', 85665, None, 85665, None, '2019-05-18 17:57:40', '2019-05-18 17:57:40', None, None, 0, 0, 0, None, None),
 (427920, 1, 0, 0, '2019-09-19 16:47:30', 0, 11, '<p>Based on <a href="http://www-bcf.usc.edu/~gareth/ISL/" rel="nofollow noreferrer">An Introduction to Statistical Learning</a>, Variance can be esti ... (640 characters truncated) ... //i.stack.imgur.com/yxDG1.png" rel="nofollow noreferrer"><img src="https://i.stack.imgur.com/yxDG1.png" alt="enter image description here"></a></p>\n', 260312, None, 0, None, None, '2019-09-19 16:47:30', "RSE's effectiveness of estimating variance in single linear regression", '<regression><linear>', 0, 0, 0, None, None),
 (396449, 2, 0, 396440, '2019-03-08 20:40:25', 0, 0, '<p>Why do you want to do Variable selection after/before/together with Regularization? Lasso also find with lambda the best number of features to use in order to simplify your model and avoiding the overfitting.</p>\n', 238550, None, 0, None, None, '2019-03-08 20:40:25', None, None, 0, 0, 0, None, None),
 (385201, 2, 0, 304742, '2019-01-01 20:42:20', 1, 0, '<p>From exchangeability of <span class="math-container">$X=(X_1, \\dotsc,X_n)$</span> it follows that all the <span class="math-container">$X_i$</spa ... (211 characters truncated) ...  joint distributios <span class="math-container">$(X_i,X_j), i\\not=j$</span> must be equal, so the off-diagonal elements must all be equal.   </p>\n', 11887, None, 0, None, None, '2019-01-01 20:42:20', None, None, 0, 0, 0, None, None)]

# Task 2: Create Single Table Queries


## Lifecycle Stage: Analyze
We can now start the analysis with our single-table queries.  First we need to create a new computed column to help with one of the queries.  The code below creates a column called LEN_BODY which is the length of the BODY text:


In [53]:
%%sql
ALTER TABLE POSTS ADD COLUMN LEN_BODY INT

 * sqlite:///chatdata.db
Done.


[]

In [54]:
%%sql
UPDATE POSTS SET LEN_BODY = LENGTH(BODY)

 * sqlite:///chatdata.db
42234 rows affected.


[]

### Single Table Queries
From the Template SQL Queries there are a series of queries requested. You are responsible for coding the result. Use the pattern above to accomplish the coding objective.  Copy and paste the question (column B Action) into a markdown cell and answer the query. Use the pattern that you have been using above. 

<div style="background-color:lemonchiffon; border:1px solid; padding:5px; margin:5px; font-weight:bold"><font color="crimson">
TODO: Enter your code below.  Just keep on inserting cells as you need them.
    </font></div>

Q1:How many posts have 0 comments?

In [55]:
%%sql
SELECT COUNT(*) FROM posts WHERE CommentCount= 0;

 * sqlite:///chatdata.db
Done.


[(21713,)]

In [56]:
sql4 = """
  SELECT COUNT(*) FROM posts WHERE CommentCount= 0;
    """
store_query("Single Table Queries", "How many posts have 0 comments?", sql4)

Q2: How many posts have 1 comments?

In [57]:
%%sql
SELECT COUNT(*) FROM posts WHERE CommentCount= 1;

 * sqlite:///chatdata.db
Done.


[(6460,)]

In [58]:
sql5 = """
SELECT COUNT(*) FROM posts WHERE CommentCount= 1;
"""
store_query("Single Table Queries", "How many posts have 1 comments?", sql5)

Q3: How many posts have 2 or more comments?

In [59]:
%%sql
SELECT COUNT(*) FROM posts WHERE CommentCount >= 2;

 * sqlite:///chatdata.db
Done.


[(14061,)]

In [60]:
sql6 = """
SELECT COUNT(*) FROM posts WHERE CommentCount >= 2;
    """
store_query("Single Table Queries", "How many posts have 2 or more comments?", sql6)

Q4:Find the 5 posts with the highest viewcount

In [61]:
sql = """
SELECT Id, SUM(viewcount) AS Highestviews
    FROM posts
        GROUP BY Id
            ORDER BY Highestviews DESC
                LIMIT 5
                """
result4 = pd.read_sql(sql, con) # con is the connection to the database
result4

Id  Highestviews
0  388566         19542
1  394118         16317
2  431370         11723
3  398646          9850
4  434128          6718

In [62]:
sql7 = """
SELECT Id, SUM(viewcount) AS Highestviews
    FROM posts
        GROUP BY Id
            ORDER BY Highestviews DESC
                LIMIT 5
                """
store_query("Single Table Queries", "Find the 5 posts with the highest viewcount", sql7)

Q5:Find the top 5 posts with the highest scores

In [63]:
sql = """
SELECT Id, SUM(Score) AS Highestscore
    FROM posts
        GROUP BY Id
            ORDER BY Highestscore DESC
                LIMIT 5
                """
result5 = pd.read_sql(sql, con) # con is the connection to the database
result5

Id  Highestscore
0  431397           101
1  394128           100
2  426878            93
3  388578            80
4  431370            77

In [64]:
sql23 = """
SELECT Id, SUM(Score) AS Highestscore
    FROM posts
        GROUP BY Id
            ORDER BY Highestscore DESC
                LIMIT 5"""
store_query("Single Table Queries", "Find the top 5 posts with the highest scores", sql23)

Q6:What are the 5 most frequent scores on posts?

In [65]:
%%sql
SELECT Score, COUNT(Score) AS frequent_scores
FROM posts
GROUP BY Score
ORDER BY frequent_scores DESC
LIMIT 5;

 * sqlite:///chatdata.db
Done.


[(0, 19888), (1, 11867), (2, 5094), (3, 2228), (4, 1059)]

In [66]:
sql8 = """
SELECT Score, COUNT(Score) AS frequent_scores
FROM posts
GROUP BY Score
ORDER BY frequent_scores DESC
LIMIT 5;"""
store_query("Single Table Queries", "What are the 5 most frequent scores on posts?", sql8)

Q7: How many posts have the keyword "data" in their tags?

In [67]:
%%sql
SELECT COUNT(*) AS posts_with_data
FROM posts
WHERE Tags LIKE '%data%';

 * sqlite:///chatdata.db
Done.


[(2242,)]

In [68]:
sql9 = """
SELECT COUNT(*) AS posts_with_data
FROM posts
WHERE Tags LIKE '%data%';
"""
store_query("Single Table Queries", "How many posts have the keyword "'data'" in their tags?", sql9)

Q8:What are the 5 most frequent commentcount for posts?

In [69]:
%%sql
SELECT CommentCount, COUNT(*) AS five_most_frequent
FROM posts
GROUP BY CommentCount
ORDER BY five_most_frequent DESC
LIMIT 5;

 * sqlite:///chatdata.db
Done.


[(0, 21713), (1, 6460), (2, 4966), (3, 3063), (4, 2026)]

In [70]:
sql10 = """
SELECT CommentCount, COUNT(*) AS five_most_frequent
FROM posts
GROUP BY CommentCount
ORDER BY five_most_frequent DESC
LIMIT 5;
"""
store_query("Single Table Queries", "What are the 5 most frequent commentcount for posts?", sql10)

Q9:How many posts have an accepted answer?
a post with no accepted answer has a value of 0 for acceptedanswerid

In [71]:
%%sql
SELECT COUNT(*) AS accepted_answers
FROM posts
WHERE AcceptedAnswerId <> 0;

 * sqlite:///chatdata.db
Done.


[(5341,)]

In [72]:
sql11 = """
SELECT COUNT(*) AS accepted_answers
FROM posts
WHERE AcceptedAnswerId <> 0;
"""
store_query("Single Table Queries", "How many posts have an accepted answer? (A post with no accepted answer has a value of 0 for acceptedanswerid)", sql11)

Q10:What is the average reputation of table users?

In [73]:
%%sql
SELECT AVG(Reputation) FROM users

 * sqlite:///chatdata.db
Done.


[(312.3509124484032,)]

In [74]:
sql12 = """
SELECT AVG(Reputation) FROM users
"""
store_query("Single Table Queries", "What is the average reputation of table users?", sql12)

Q11:What are the min and max reputation of users?

In [75]:
%%sql
SELECT MAX(Reputation) FROM users;

 * sqlite:///chatdata.db
Done.


[(228662,)]

In [76]:
%%sql
SELECT MIN(Reputation) FROM users;

 * sqlite:///chatdata.db
Done.


[(1,)]

In [77]:
sql13 = """
SELECT MIN(Reputation) FROM users
SELECT MAX(Reputation) FROM users;
"""
store_query("Single Table Queries", "What are the min and max reputation of users? (split into two cells)", sql13)

Q12:What is the length of the body of 5 most viewed posts?

In [78]:
%%sql
SELECT Title, ViewCount as View_Count, LENGTH(Body) as Number_of_Characters
FROM posts
GROUP BY Title
ORDER BY View_Count DESC
LIMIT 5;

 * sqlite:///chatdata.db
Done.


[('Is it wrong to rephrase "1 in 80 deaths is caused by a car accident" as "1 in 80 people die as a result of a car accident?"', 19542, 2270),
 ('Why do neural networks need so many training examples to perform?', 16317, 512),
 ('Is there a name for the phenomenon of false positives counterintuitively outstripping true positives', 11723, 811),
 ('What does "Scientists rise up against statistical significance" mean? (Comment in Nature)', 9850, 2148),
 ('Famous statistical wins and horror stories for teaching purposes', 6718, 1172)]

In [79]:
sql22 = """
SELECT Title, ViewCount as View_Count, LENGTH(Body) as Number_of_Characters
FROM posts
GROUP BY Title
ORDER BY View_Count DESC
LIMIT 5;
"""
store_query("Single Table Queries", "What is the length of the body of 5 most viewed posts?", sql22)

Q13:How many different locations are there in the users table?

In [80]:
%%sql
SELECT COUNT(DISTINCT Location) FROM users;

 * sqlite:///chatdata.db
Done.


[(1900,)]

In [81]:
sql14 = """
SELECT COUNT(DISTINCT Location) FROM users;
"""
store_query("Single Table Queries", "How many different locations are there in the users table?", sql14)

Q14:What are the top 5 locations of users?

In [82]:
%%sql
SELECT Location, COUNT(Location) Top_Locations
FROM users
GROUP BY Location
ORDER BY Top_Locations DESC
LIMIT 5;

 * sqlite:///chatdata.db
Done.


[('Germany', 117),
 ('India', 100),
 ('United States', 69),
 ('Paris, France', 66),
 ('London, United Kingdom', 63)]

In [83]:
sql15 = """
SELECT Location, COUNT(Location) Top_Locations
FROM users
GROUP BY Location
ORDER BY Top_Locations DESC
LIMIT 5;
"""
store_query("Single Table Queries", "What are the top 5 locations of users?", sql15)

Q15:Rank the days of the week from highest to lowest in terms of the volume of ViewCount as a percentage.

In [84]:
%%sql
SELECT ROUND(SUM(ViewCount * 100.0)  /
     (SELECT SUM(ViewCount) FROM Posts), 2) AS View_Count_Percent,
    CASE strftime('%w', CreationDate)
        WHEN '0' THEN 'Sunday'
        WHEN '1' THEN 'Monday'
        WHEN '2' THEN 'Tuesday'
        WHEN '3' THEN 'Wednesday'
        WHEN '4' THEN 'Thursday'
        WHEN '5' THEN 'Friday'
        WHEN '6' THEN 'Saturday'
        ELSE 'Unknown'
    END AS Day_of_Week
FROM posts
GROUP BY Day_of_Week
ORDER BY View_Count_Percent DESC;

 * sqlite:///chatdata.db
Done.


[(16.82, 'Thursday'),
 (16.77, 'Wednesday'),
 (16.26, 'Tuesday'),
 (15.81, 'Monday'),
 (13.56, 'Friday'),
 (11.9, 'Sunday'),
 (8.89, 'Saturday')]

In [85]:
sql16 = """
SELECT ROUND(SUM(ViewCount * 100.0)  /
     (SELECT SUM(ViewCount) FROM Posts), 2) AS View_Count_Percent,
    CASE strftime('%w', CreationDate)
        WHEN '0' THEN 'Sunday'
        WHEN '1' THEN 'Monday'
        WHEN '2' THEN 'Tuesday'
        WHEN '3' THEN 'Wednesday'
        WHEN '4' THEN 'Thursday'
        WHEN '5' THEN 'Friday'
        WHEN '6' THEN 'Saturday'
        ELSE 'Unknown'
    END AS Day_of_Week
FROM posts
GROUP BY Day_of_Week
ORDER BY View_Count_Percent DESC;
"""
store_query("Single Table Queries", "Rank the days of the week from highest to lowest in terms of the volume of ViewCount as a percentage.", sql16)

# Task 3: Cross Table Queries

## Lifecycle Stage: Analyze
Let's continue the analysis with our multi-table queries.  

### Cross Table Queries

From the Template SQL Queries there are a series of queries requested. You are responsible for coding the result. Use the pattern above to accomplish the coding objective.  Copy and paste the question (column B Action) into a markdown cell and answer the query. Use the pattern that you have been using above. 

<div style="background-color:lemonchiffon; border:1px solid; padding:5px; margin:5px; font-weight:bold"><font color="crimson">
TODO: Enter your code below.  Just keep on inserting cells as you need them.
    </font></div>



Q:16 How many posts have been created by a user that has a filled out the "AboutMe" section?

In [86]:
%%sql
SELECT Count(*) AS number_of_posts_with_about_me
FROM posts
JOIN users ON users.Id = posts.OwnerUserId
WHERE users.AboutMe <> '';

 * sqlite:///chatdata.db
Done.


[(17189,)]

In [87]:
sql17 = """
SELECT Count(*) AS number_of_posts_with_about_me
FROM posts
JOIN users ON users.Id = posts.OwnerUserId
WHERE users.AboutMe <> '';
"""
store_query("Cross Table Queries", "How many posts have been created by a user that has a filled out the ""AboutMe"" section?", sql17)

Q17: Considering only the users with an "AboutMe," how many posts are there per user?*

In [88]:
%%sql
SELECT COUNT(*) / (SELECT CAST(COUNT(*) AS FLOAT) FROM USERS WHERE ABOUTME !='') AS posts_per_user
    FROM POSTS P
    JOIN USERS U ON P.OWNERUSERID = U.ID
    WHERE U.ABOUTME !=''

 * sqlite:///chatdata.db
Done.


[(4.274807261875155,)]

In [89]:
sql18 = """
SELECT COUNT(*) / (SELECT CAST(COUNT(*) AS FLOAT) FROM USERS WHERE ABOUTME !='') AS posts_per_user
    FROM POSTS P
    JOIN USERS U ON P.OWNERUSERID = U.ID
    WHERE U.ABOUTME !=''
"""
store_query("Cross Table Queries", "Considering only the users with an ""AboutMe,"" how many posts are there per user?", sql18)

4.27

Q18: Not taking into account the commentcount field in the table posts, what are the Top 10 posts in terms of number of comments?

In [90]:
%%sql
SELECT posts.Id as Post_Number, Count(DISTINCT comments.Id) as Number_of_Comments
FROM posts
INNER JOIN comments ON posts.Id = comments.postId
Group by posts.Id
Order by Number_of_comments DESC
Limit 10;

 * sqlite:///chatdata.db
Done.


[(386853, 66),
 (386556, 34),
 (395232, 31),
 (418910, 31),
 (402987, 27),
 (386075, 26),
 (394118, 24),
 (398828, 23),
 (402950, 23),
 (396111, 22)]

In [91]:
sql19 = """
SELECT posts.Id as Post_Number, Count(DISTINCT comments.Id) as Number_of_Comments
FROM posts
INNER JOIN comments ON posts.Id = comments.postId
Group by posts.Id
Order by Number_of_comments DESC
Limit 10;
"""
store_query("Cross Table Queries", "Not taking into account the commentcount field in the table posts, what are the Top 10 posts in terms of number of comments?", sql19)

Q19: What are the Top 10 posts which have the highest cummulative (post score + comment score) score? 

In [92]:
%%sql
SELECT posts.Id as Post_Id, (sum(comments.score) + posts.score) as Cummulative_score 
FROM posts
INNER JOIN comments ON posts.Id = comments.postId
GROUP BY posts.Id
ORDER BY Cummulative_score DESC
Limit 10;

 * sqlite:///chatdata.db
Done.


[(394118, 306),
 (394128, 169),
 (388578, 141),
 (398653, 111),
 (388566, 101),
 (398646, 99),
 (421677, 96),
 (420526, 92),
 (400317, 72),
 (388582, 69)]

394

In [93]:
sql20 = """
SELECT posts.Id as Post_Id, (sum(comments.score) + posts.score) as Cummulative_score 
FROM posts
INNER JOIN comments ON posts.Id = comments.postId
GROUP BY posts.Id
ORDER BY Cummulative_score DESC
Limit 10;
"""
store_query("Cross Table Queries", "What are the Top 10 posts which have the highest cummulative (post score + comment score) score?", sql20)

Q20: Who are the top 10 users who comment the most? 

In [94]:
%%sql
SELECT users.DisplayName as User, users.Id as User_Id, users.Reputation as Reputation, posts.CommentCount as Number_of_Comments
FROM users
INNER JOIN comments ON comments.UserId = users.Id
INNER JOIN posts ON comments.postId = posts.Id
GROUP BY user
ORDER BY users.Reputation DESC
LIMIT 10;

 * sqlite:///chatdata.db
Done.


[('Glen_b -Reinstate Monica', 805, 228662, 1),
 ('whuber', 919, 223056, 2),
 ('gung - Reinstate Monica', 7290, 115531, 1),
 ('Peter Flom - Reinstate Monica', 686, 85077, 2),
 ('amoeba says Reinstate Monica', 28666, 75024, 2),
 ('Tim', 35989, 71548, 2),
 ("Xi'an", 7224, 65999, 3),
 ('Frank Harrell', 4253, 59952, 3),
 ('S. Kolassa - Reinstate Monica', 1352, 59160, 2),
 ('Sycorax says Reinstate Monica', 22311, 51155, 2)]

In [95]:
sql21 = """
SELECT users.DisplayName as User, users.Id as User_Id, users.Reputation as Reputation, posts.CommentCount as Number_of_Comments
FROM users
INNER JOIN comments ON comments.UserId = users.Id
INNER JOIN posts ON comments.postId = posts.Id
GROUP BY user
ORDER BY users.Reputation DESC
LIMIT 10;
"""
store_query("Cross Table Queries", "Who are the top 10 users who comment the most?", sql21)

Q21: Who are the top 10 users who post the most?  

In [96]:
%%sql
SELECT users.DisplayName as User, users.Id as User_Id, users.Reputation as Reputation, Count(DISTINCT Posts.Id) as Number_of_posts
FROM users
INNER JOIN comments ON comments.UserId = users.Id
INNER JOIN posts ON comments.postId = posts.Id
GROUP BY user
ORDER BY users.Reputation DESC
LIMIT 10;

 * sqlite:///chatdata.db
Done.


[('Glen_b -Reinstate Monica', 805, 228662, 719),
 ('whuber', 919, 223056, 2044),
 ('gung - Reinstate Monica', 7290, 115531, 241),
 ('Peter Flom - Reinstate Monica', 686, 85077, 257),
 ('amoeba says Reinstate Monica', 28666, 75024, 70),
 ('Tim', 35989, 71548, 284),
 ("Xi'an", 7224, 65999, 320),
 ('Frank Harrell', 4253, 59952, 153),
 ('S. Kolassa - Reinstate Monica', 1352, 59160, 261),
 ('Sycorax says Reinstate Monica', 22311, 51155, 284)]

In [97]:
sql22 = """
SELECT users.DisplayName as User, users.Id as User_Id, users.Reputation as Reputation, Count(DISTINCT Posts.Id) as Number_of_posts
FROM users
INNER JOIN comments ON comments.UserId = users.Id
INNER JOIN posts ON comments.postId = posts.Id
GROUP BY user
ORDER BY users.Reputation DESC
LIMIT 10;
"""
store_query("Cross Table Queries", "Who are the top 10 users who post the most?", sql22)

In [98]:
# Prove it works
%sql SELECT * FROM queries

 * sqlite:///chatdata.db
Done.


[('Single Table Queries', 'Which 5 users have viewed the most times and what is the sum of those views per user?', '\nSELECT Id, SUM(Views) AS TotalViews\n    FROM Users\n        GROUP BY Id\n            ORDER BY TotalViews DESC\n                LIMIT 5\n                '),
 ('Creating Tables', 'Create table comments', '\n    CREATE TABLE "comments" (\n    "Id" INTEGER,\n    "PostId" INTEGER,\n    "Score" INTEGER,\n    "Text" TEXT,\n    "CreationDate" TEXT,\n    "UserId" INTEGER\n    )\n    '),
 ('Creating Tables', 'Create table posts', '\n    CREATE TABLE "posts" (\n    "Id" INTEGER NOT NULL PRIMARY KEY,\n    "PostTypeId" INTEGER,\n    "AcceptedAnswerId" INTEGER,\n    "ParentId" INTE ... (321 characters truncated) ... "AnswerCount" INTEGER,\n    "CommentCount" INTEGER,\n    "FavoriteCount" INTEGER,\n    "ClosedDate" TEXT,\n    "CommunityOwnedDate" TEXT\n    )\n    '),
 ('Creating Tables', 'Create table users', '\n    CREATE TABLE "users" (\n    "Id" INTEGER NOT NULL PRIMARY KEY,\n    "Reputation" INTEGER,\n    "CreationDate" TEXT,\n    "DisplayName" TEXT,\n  ... (83 characters truncated) ... Me" TEXT,\n    "Views" INTEGER,\n    "UpVotes" INTEGER,\n    "DownVotes" INTEGER,\n    "ProfileImageUrl" TEXT,\n    "AccountId" INTEGER\n    );\n    '),
 ('Single Table Queries', 'How many posts have 0 comments?', '\n  SELECT COUNT(*) FROM posts WHERE CommentCount= 0;\n    '),
 ('Single Table Queries', 'How many posts have 1 comments?', '\nSELECT COUNT(*) FROM posts WHERE CommentCount= 1;\n'),
 ('Single Table Queries', 'How many posts have 2 or more comments?', '\nSELECT COUNT(*) FROM posts WHERE CommentCount >= 2;\n    '),
 ('Single Table Queries', 'Find the 5 posts with the highest viewcount', '\nSELECT Id, SUM(viewcount) AS Highestviews\n    FROM posts\n        GROUP BY Id\n            ORDER BY Highestviews DESC\n                LIMIT 5\n                '),
 ('Single Table Queries', 'Find the top 5 posts with the highest scores', '\nSELECT Id, SUM(Score) AS Highestscore\n    FROM posts\n        GROUP BY Id\n            ORDER BY Highestscore DESC\n                LIMIT 5'),
 ('Single Table Queries', 'What are the 5 most frequent scores on posts?', '\nSELECT Score, COUNT(Score) AS frequent_scores\nFROM posts\nGROUP BY Score\nORDER BY frequent_scores DESC\nLIMIT 5;'),
 ('Single Table Queries', 'How many posts have the keyword data in their tags?', "\nSELECT COUNT(*) AS posts_with_data\nFROM posts\nWHERE Tags LIKE '%data%';\n"),
 ('Single Table Queries', 'What are the 5 most frequent commentcount for posts?', '\nSELECT CommentCount, COUNT(*) AS five_most_frequent\nFROM posts\nGROUP BY CommentCount\nORDER BY five_most_frequent DESC\nLIMIT 5;\n'),
 ('Single Table Queries', 'How many posts have an accepted answer? (A post with no accepted answer has a value of 0 for acceptedanswerid)', '\nSELECT COUNT(*) AS accepted_answers\nFROM posts\nWHERE AcceptedAnswerId <> 0;\n'),
 ('Single Table Queries', 'What is the average reputation of table users?', '\nSELECT AVG(Reputation) FROM users\n'),
 ('Single Table Queries', 'What are the min and max reputation of users? (split into two cells)', '\nSELECT MIN(Reputation) FROM users\nSELECT MAX(Reputation) FROM users;\n'),
 ('Single Table Queries', 'What is the length of the body of 5 most viewed posts?', '\nSELECT Title, ViewCount as View_Count, LENGTH(Body) as Number_of_Characters\nFROM posts\nGROUP BY Title\nORDER BY View_Count DESC\nLIMIT 5;\n'),
 ('Single Table Queries', 'How many different locations are there in the users table?', '\nSELECT COUNT(DISTINCT Location) FROM users;\n'),
 ('Single Table Queries', 'What are the top 5 locations of users?', '\nSELECT Location, COUNT(Location) Top_Locations\nFROM users\nGROUP BY Location\nORDER BY Top_Locations DESC\nLIMIT 5;\n'),
 ('Single Table Queries', 'Rank the days of the week from highest to lowest in terms of the volume of ViewCount as a percentage.', "\nSELECT ROUND(SUM(ViewCount * 100.0)  /\n     (SELECT SUM(ViewCount) FROM Posts), 2) AS View_Count_Pe

# Task 4: Check the Queries Table

Now let's tidy up and check the queries table.

First let's check it's contents:

In [99]:
%sql SELECT * FROM queries

 * sqlite:///chatdata.db
Done.


[('Single Table Queries', 'Which 5 users have viewed the most times and what is the sum of those views per user?', '\nSELECT Id, SUM(Views) AS TotalViews\n    FROM Users\n        GROUP BY Id\n            ORDER BY TotalViews DESC\n                LIMIT 5\n                '),
 ('Creating Tables', 'Create table comments', '\n    CREATE TABLE "comments" (\n    "Id" INTEGER,\n    "PostId" INTEGER,\n    "Score" INTEGER,\n    "Text" TEXT,\n    "CreationDate" TEXT,\n    "UserId" INTEGER\n    )\n    '),
 ('Creating Tables', 'Create table posts', '\n    CREATE TABLE "posts" (\n    "Id" INTEGER NOT NULL PRIMARY KEY,\n    "PostTypeId" INTEGER,\n    "AcceptedAnswerId" INTEGER,\n    "ParentId" INTE ... (321 characters truncated) ... "AnswerCount" INTEGER,\n    "CommentCount" INTEGER,\n    "FavoriteCount" INTEGER,\n    "ClosedDate" TEXT,\n    "CommunityOwnedDate" TEXT\n    )\n    '),
 ('Creating Tables', 'Create table users', '\n    CREATE TABLE "users" (\n    "Id" INTEGER NOT NULL PRIMARY KEY,\n    "Reputation" INTEGER,\n    "CreationDate" TEXT,\n    "DisplayName" TEXT,\n  ... (83 characters truncated) ... Me" TEXT,\n    "Views" INTEGER,\n    "UpVotes" INTEGER,\n    "DownVotes" INTEGER,\n    "ProfileImageUrl" TEXT,\n    "AccountId" INTEGER\n    );\n    '),
 ('Single Table Queries', 'How many posts have 0 comments?', '\n  SELECT COUNT(*) FROM posts WHERE CommentCount= 0;\n    '),
 ('Single Table Queries', 'How many posts have 1 comments?', '\nSELECT COUNT(*) FROM posts WHERE CommentCount= 1;\n'),
 ('Single Table Queries', 'How many posts have 2 or more comments?', '\nSELECT COUNT(*) FROM posts WHERE CommentCount >= 2;\n    '),
 ('Single Table Queries', 'Find the 5 posts with the highest viewcount', '\nSELECT Id, SUM(viewcount) AS Highestviews\n    FROM posts\n        GROUP BY Id\n            ORDER BY Highestviews DESC\n                LIMIT 5\n                '),
 ('Single Table Queries', 'Find the top 5 posts with the highest scores', '\nSELECT Id, SUM(Score) AS Highestscore\n    FROM posts\n        GROUP BY Id\n            ORDER BY Highestscore DESC\n                LIMIT 5'),
 ('Single Table Queries', 'What are the 5 most frequent scores on posts?', '\nSELECT Score, COUNT(Score) AS frequent_scores\nFROM posts\nGROUP BY Score\nORDER BY frequent_scores DESC\nLIMIT 5;'),
 ('Single Table Queries', 'How many posts have the keyword data in their tags?', "\nSELECT COUNT(*) AS posts_with_data\nFROM posts\nWHERE Tags LIKE '%data%';\n"),
 ('Single Table Queries', 'What are the 5 most frequent commentcount for posts?', '\nSELECT CommentCount, COUNT(*) AS five_most_frequent\nFROM posts\nGROUP BY CommentCount\nORDER BY five_most_frequent DESC\nLIMIT 5;\n'),
 ('Single Table Queries', 'How many posts have an accepted answer? (A post with no accepted answer has a value of 0 for acceptedanswerid)', '\nSELECT COUNT(*) AS accepted_answers\nFROM posts\nWHERE AcceptedAnswerId <> 0;\n'),
 ('Single Table Queries', 'What is the average reputation of table users?', '\nSELECT AVG(Reputation) FROM users\n'),
 ('Single Table Queries', 'What are the min and max reputation of users? (split into two cells)', '\nSELECT MIN(Reputation) FROM users\nSELECT MAX(Reputation) FROM users;\n'),
 ('Single Table Queries', 'What is the length of the body of 5 most viewed posts?', '\nSELECT Title, ViewCount as View_Count, LENGTH(Body) as Number_of_Characters\nFROM posts\nGROUP BY Title\nORDER BY View_Count DESC\nLIMIT 5;\n'),
 ('Single Table Queries', 'How many different locations are there in the users table?', '\nSELECT COUNT(DISTINCT Location) FROM users;\n'),
 ('Single Table Queries', 'What are the top 5 locations of users?', '\nSELECT Location, COUNT(Location) Top_Locations\nFROM users\nGROUP BY Location\nORDER BY Top_Locations DESC\nLIMIT 5;\n'),
 ('Single Table Queries', 'Rank the days of the week from highest to lowest in terms of the volume of ViewCount as a percentage.', "\nSELECT ROUND(SUM(ViewCount * 100.0)  /\n     (SELECT SUM(ViewCount) FROM Posts), 2) AS View_Count_Pe

## Drop Duplicates

You likely have some duplicates. Lets drop them. 

In [100]:
# Read the queries table into pandas
sql = 'SELECT * FROM queries'
queries = pd.read_sql(sql, con)

# Drop duplicates
queries.drop_duplicates(inplace = True) # drop duplicates

## Case Issues

Remember, SQL is case insensitive. Pandas IS case sensitive. Lets deal with this now by making all of the text uppercase. 

In [101]:
for col in queries.columns:
    queries[col] = queries[col].str.upper()
    
queries

task                                             action  \
0   SINGLE TABLE QUERIES  WHICH 5 USERS HAVE VIEWED THE MOST TIMES AND W...   
1        CREATING TABLES                              CREATE TABLE COMMENTS   
2        CREATING TABLES                                 CREATE TABLE POSTS   
3        CREATING TABLES                                 CREATE TABLE USERS   
4   SINGLE TABLE QUERIES                    HOW MANY POSTS HAVE 0 COMMENTS?   
5   SINGLE TABLE QUERIES                    HOW MANY POSTS HAVE 1 COMMENTS?   
6   SINGLE TABLE QUERIES            HOW MANY POSTS HAVE 2 OR MORE COMMENTS?   
7   SINGLE TABLE QUERIES        FIND THE 5 POSTS WITH THE HIGHEST VIEWCOUNT   
8   SINGLE TABLE QUERIES       FIND THE TOP 5 POSTS WITH THE HIGHEST SCORES   
9   SINGLE TABLE QUERIES      WHAT ARE THE 5 MOST FREQUENT SCORES ON POSTS?   
10  SINGLE TABLE QUERIES  HOW MANY POSTS HAVE THE KEYWORD DATA IN THEIR ...   
11  SINGLE TABLE QUERIES  WHAT ARE THE 5 MOST FREQUENT COMMENTCOUNT FOR ...   
12  SINGLE TABLE QUERIES  HOW MANY POSTS HAVE AN ACCEPTED ANSWER? (A POS...   
13  SINGLE TABLE QUERIES     WHAT IS THE AVERAGE REPUTATION OF TABLE USERS?   
14  SINGLE TABLE QUERIES  WHAT ARE THE MIN AND MAX REPUTATION OF USERS? ...   
15  SINGLE TABLE QUERIES  WHAT IS THE LENGTH OF THE BODY OF 5 MOST VIEWE...   
16  SINGLE TABLE QUERIES  HOW MANY DIFFERENT LOCATIONS ARE THERE IN THE ...   
17  SINGLE TABLE QUERIES             WHAT ARE THE TOP 5 LOCATIONS OF USERS?   
18  SINGLE TABLE QUERIES  RANK THE DAYS OF THE WEEK FROM HIGHEST TO LOWE...   
19   CROSS TABLE QUERIES  HOW MANY POSTS HAVE BEEN CREATED BY A USER THA...   
20   CROSS TABLE QUERIES  CONSIDERING ONLY THE USERS WITH AN ABOUTME, HO...   
21   CROSS TABLE QUERIES  NOT TAKING INTO ACCOUNT THE COMMENTCOUNT FIELD...   
22   CROSS TABLE QUERIES  WHAT ARE THE TOP 10 POSTS WHICH HAVE THE HIGHE...   
23   CROSS TABLE QUERIES     WHO ARE THE TOP 10 USERS WHO COMMENT THE MOST?   
24   CROSS TABLE QUERIES        WHO ARE THE TOP 10 USERS WHO POST THE MOST?   

                                                query  
0   \nSELECT ID, SUM(VIEWS) AS TOTALVIEWS\n    FRO...  
1   \n    CREATE TABLE "COMMENTS" (\n    "ID" INTE...  
2   \n    CREATE TABLE "POSTS" (\n    "ID" INTEGER...  
3   \n    CREATE TABLE "USERS" (\n    "ID" INTEGER...  
4   \n  SELECT COUNT(*) FROM POSTS WHERE COMMENTCO...  
5   \nSELECT COUNT(*) FROM POSTS WHERE COMMENTCOUN...  
6   \nSELECT COUNT(*) FROM POSTS WHERE COMMENTCOUN...  
7   \nSELECT ID, SUM(VIEWCOUNT) AS HIGHESTVIEWS\n ...  
8   \nSELECT ID, SUM(SCORE) AS HIGHESTSCORE\n    F...  
9   \nSELECT SCORE, COUNT(SCORE) AS FREQUENT_SCORE...  
10  \nSELECT COUNT(*) AS POSTS_WITH_DATA\nFROM POS...  
11  \nSELECT COMMENTCOUNT, COUNT(*) AS FIVE_MOST_F...  
12  \nSELECT COUNT(*) AS ACCEPTED_ANSWERS\nFROM PO...  
13              \nSELECT AVG(REPUTATION) FROM USERS\n  
14  \nSELECT MIN(REPUTATION) FROM USERS\nSELECT MA...  
15  \nSELECT TITLE, VIEWCOUNT AS VIEW_COUNT, LENGT...  
16    \nSELECT COUNT(DISTINCT LOCATION) FROM USERS;\n  
17  \nSELECT LOCATION, COUNT(LOCATION) TOP_LOCATIO...  
18  \nSELECT ROUND(SUM(VIEWCOUNT * 100.0)  /\n    ...  
19  \nSELECT COUNT(*) AS NUMBER_OF_POSTS_WITH_ABOU...  
20  \nSELECT COUNT(*) / (SELECT CAST(COUNT(*) AS F...  
21  \nSELECT POSTS.ID AS POST_NUMBER, COUNT(DISTIN...  
22  \nSELECT POSTS.ID AS POST_ID, (SUM(COMMENTS.SC...  
23  \nSELECT USERS.DISPLAYNAME AS USER, USERS.ID A...  
24  \nSELECT USERS.DISPLAYNAME AS USER, USERS.ID A...

In [102]:
# Write the now deduped uppercase dataframe back to sqlite and replace the table
queries.to_sql('queries', con, if_exists='replace', index=False)

## Use Case

Now that we have this queries table, lets give you some ideas about how you would use it.

Suppose you wanted to find all of the queries where you did a GROUP BY:

In [103]:
%%sql
SELECT query
    FROM queries
        WHERE query LIKE '%GROUP BY%'

 * sqlite:///chatdata.db
Done.


[('\nSELECT ID, SUM(VIEWS) AS TOTALVIEWS\n    FROM USERS\n        GROUP BY ID\n            ORDER BY TOTALVIEWS DESC\n                LIMIT 5\n                ',),
 ('\nSELECT ID, SUM(VIEWCOUNT) AS HIGHESTVIEWS\n    FROM POSTS\n        GROUP BY ID\n            ORDER BY HIGHESTVIEWS DESC\n                LIMIT 5\n                ',),
 ('\nSELECT ID, SUM(SCORE) AS HIGHESTSCORE\n    FROM POSTS\n        GROUP BY ID\n            ORDER BY HIGHESTSCORE DESC\n                LIMIT 5',),
 ('\nSELECT SCORE, COUNT(SCORE) AS FREQUENT_SCORES\nFROM POSTS\nGROUP BY SCORE\nORDER BY FREQUENT_SCORES DESC\nLIMIT 5;',),
 ('\nSELECT COMMENTCOUNT, COUNT(*) AS FIVE_MOST_FREQUENT\nFROM POSTS\nGROUP BY COMMENTCOUNT\nORDER BY FIVE_MOST_FREQUENT DESC\nLIMIT 5;\n',),
 ('\nSELECT TITLE, VIEWCOUNT AS VIEW_COUNT, LENGTH(BODY) AS NUMBER_OF_CHARACTERS\nFROM POSTS\nGROUP BY TITLE\nORDER BY VIEW_COUNT DESC\nLIMIT 5;\n',),
 ('\nSELECT LOCATION, COUNT(LOCATION) TOP_LOCATIONS\nFROM USERS\nGROUP BY LOCATION\nORDER BY TOP_LOCATIONS DESC\nLIMIT 5;\n',),
 ("\nSELECT ROUND(SUM(VIEWCOUNT * 100.0)  /\n     (SELECT SUM(VIEWCOUNT) FROM POSTS), 2) AS VIEW_COUNT_PERCENT,\n    CASE STRFTIME('%W', CREATIONDATE)\n ... (200 characters truncated) ...       WHEN '6' THEN 'SATURDAY'\n        ELSE 'UNKNOWN'\n    END AS DAY_OF_WEEK\nFROM POSTS\nGROUP BY DAY_OF_WEEK\nORDER BY VIEW_COUNT_PERCENT DESC;\n",),
 ('\nSELECT POSTS.ID AS POST_NUMBER, COUNT(DISTINCT COMMENTS.ID) AS NUMBER_OF_COMMENTS\nFROM POSTS\nINNER JOIN COMMENTS ON POSTS.ID = COMMENTS.POSTID\nGROUP BY POSTS.ID\nORDER BY NUMBER_OF_COMMENTS DESC\nLIMIT 10;\n',),
 ('\nSELECT POSTS.ID AS POST_ID, (SUM(COMMENTS.SCORE) + POSTS.SCORE) AS CUMMULATIVE_SCORE \nFROM POSTS\nINNER JOIN COMMENTS ON POSTS.ID = COMMENTS.POSTID\nGROUP BY POSTS.ID\nORDER BY CUMMULATIVE_SCORE DESC\nLIMIT 10;\n',),
 ('\nSELECT USERS.DISPLAYNAME AS USER, USERS.ID AS USER_ID, USERS.REPUTATION AS REPUTATION, POSTS.COMMENTCOUNT AS NUMBER_OF_COMMENTS\nFROM USERS\nINNER  ... (2 characters truncated) ... IN COMMENTS ON COMMENTS.USERID = USERS.ID\nINNER JOIN POSTS ON COMMENTS.POSTID = POSTS.ID\nGROUP BY USER\nORDER BY USERS.REPUTATION DESC\nLIMIT 10;\n',),
 ('\nSELECT USERS.DISPLAYNAME AS USER, USERS.ID AS USER_ID, USERS.REPUTATION AS REPUTATION, COUNT(DISTINCT POSTS.ID) AS NUMBER_OF_POSTS\nFROM USERS\nINN ... (5 characters truncated) ... IN COMMENTS ON COMMENTS.USERID = USERS.ID\nINNER JOIN POSTS ON COMMENTS.POSTID = POSTS.ID\nGROUP BY USER\nORDER BY USERS.REPUTATION DESC\nLIMIT 10;\n',)]

## Now Your Turn
Find the queries that have 'DISTINCT' in them. You can do it with the %sql command or with Pandas and sql.

<div style="background-color:lemonchiffon; border:1px solid; padding:5px; margin:5px; font-weight:bold"><font color="crimson">
TODO: Complete the following code cell
    </font></div>

In [104]:
%%sql
SELECT query
    FROM queries
        WHERE query LIKE '%DISTINCT%'

 * sqlite:///chatdata.db
Done.


[('\nSELECT COUNT(DISTINCT LOCATION) FROM USERS;\n',),
 ('\nSELECT POSTS.ID AS POST_NUMBER, COUNT(DISTINCT COMMENTS.ID) AS NUMBER_OF_COMMENTS\nFROM POSTS\nINNER JOIN COMMENTS ON POSTS.ID = COMMENTS.POSTID\nGROUP BY POSTS.ID\nORDER BY NUMBER_OF_COMMENTS DESC\nLIMIT 10;\n',),
 ('\nSELECT USERS.DISPLAYNAME AS USER, USERS.ID AS USER_ID, USERS.REPUTATION AS REPUTATION, COUNT(DISTINCT POSTS.ID) AS NUMBER_OF_POSTS\nFROM USERS\nINN ... (5 characters truncated) ... IN COMMENTS ON COMMENTS.USERID = USERS.ID\nINNER JOIN POSTS ON COMMENTS.POSTID = POSTS.ID\nGROUP BY USER\nORDER BY USERS.REPUTATION DESC\nLIMIT 10;\n',)]

# Close SQLite

It is good practise to close all relational databases as soon as you are finished updating them.

In [105]:
con.close()

# All Done!

Great job. You now have a good idea for how to use sql and pandas with sql. You can create your own databases from csv files and you can do extensive querying using sql. These are valuable skills that will take you a long ways in todays technological world.